frontend/request.go

```diff
@@ -137,23 +142,86 @@ func marshalDockerfile(ctx context.Context, dt []byte, opts ...llb.ConstraintsOp
 	return st.Marshal(ctx)
 }
 
-func MaybeSign(ctx context.Context, client gwclient.Client, st llb.State, spec *dalec.Spec, targetKey string) (llb.State, error) {
-	if signingDisabled(client) {
-		Warn(ctx, client, st, "Signing disabled by build-arg "+keySkipSigningArg)
-		return st, nil
+func getSigningConfigFromContext(ctx context.Context, client gwclient.Client, cfgPath string, configCtxName string, sOpt dalec.SourceOpts) (*dalec.PackageSigner, error) {
+	sc := dalec.SourceContext{Name: configCtxName}
+	signConfigState, err := sc.AsState([]string{cfgPath}, nil, sOpt)
+	if err != nil {
+		return nil, err
 	}
 
-	signer := spec.GetSigner(targetKey)
-	if signer == nil {
-		return st, nil
+	scDef, err := signConfigState.Marshal(ctx)
+	if err != nil {
+		return nil, err
 	}
 
-	signed, err := forwardToSigner(ctx, client, signer, st)
+	res, err := client.Solve(ctx, gwclient.SolveRequest{
+		Definition: scDef.ToPB(),
+	})
 	if err != nil {
-		return llb.Scratch(), err
+		return nil, err
+	}
+
+	ref, err := res.SingleRef()
+	if err != nil {
+		return nil, err
+	}
+
+	dt, err := ref.ReadFile(ctx, gwclient.ReadRequest{
+		Filename: cfgPath,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	var pc dalec.PackageConfig
+	if err := yaml.Unmarshal(dt, &pc); err != nil {
+		return nil, err
+	}
+
+	return pc.Signer, nil
+}
+
+func MaybeSign(ctx context.Context, client gwclient.Client, st llb.State, spec *dalec.Spec, targetKey string, sOpt dalec.SourceOpts) (llb.State, error) {
+	if signingDisabled(client) {
+		Warnf(ctx, client, st, "Signing disabled by build-arg %q", keySkipSigningArg)
+		return st, nil
+	}
+
+	cfg, warning := spec.GetSigner(targetKey)
+	switch cfgPath := getUserSignConfigPath(client); cfgPath {
```

Comment:
Can we restructure this like to reduce indentation?

```go
cfgPath := getUserSignConfigPath(client)
if cfgPath != "" {
  // handle
  return updatedSt, nil
}

cfg, warning := spec.GetSigner(targetKey)
// handle the rest
```

Reply:
done